File: ruby/easy/merge_two_binary_trees.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} t1
# @param {TreeNode} t2
# @return {TreeNode}
def merge_trees(t1, t2)
    return nil if t1.nil? && t2.nil?
    sum = (t1.nil? ? 0 : t1.val) + (t2.nil? ? 0 : t2.val)
    res = TreeNode.new(sum)
    res.left = merge_trees((t1 ? t1.left : nil), (t2 ? t2.left : nil))
    res.right = merge_trees((t1 ? t1.right : nil), (t2 ? t2.right : nil))
    return res
end

File: java/src/leetcode/easy/MergeTwoBinaryTrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class MergeTwoBinaryTrees {
    public TreeNode mergeTrees(TreeNode t1, TreeNode t2) {
        if(t1 == null && t2 == null) return null;
        int sum = ((t1 != null) ? t1.val : 0) + ((t2 != null) ? t2.val : 0);
        TreeNode node = new TreeNode(sum);
        node.left = mergeTrees(((t1 != null) ? t1.left : null), ((t2 != null) ? t2.left : null));
        node.right = mergeTrees(((t1 != null) ? t1.right : null), ((t2 != null) ? t2.right : null));
        return node;
    }
}
